src/codesearch.cc

```diff
@@ -352,6 +368,49 @@ class searcher {
     friend class code_searcher::search_thread;
 };
 
+class filename_searcher {
+public:
+    filename_searcher(const code_searcher *cc,
+                      const query &q) :
+        cc_(cc), query_(&q), limiter_(q.max_matches) {}
+
+    void operator()(indexed_file *file);
+
+protected:
+    const code_searcher *cc_;
+    const query *query_;
+    thread_queue<file_result*> queue_;
+    atomic_int matches_;
+    search_limiter limiter_;
+
+    friend class code_searcher::search_thread;
+};
+
+void filename_searcher::operator()(indexed_file *file)
+{
+    if (limiter_.why())
```

Comment:
Hmm. I wonder if we should rename `why()`; It is intended only to be called post-facto, and `exit_early()` should be called in the loop. I think I short-circuited to checking `exit_reason_` either over concern about the  `gettimeofday` overhead (likely overblow; gtod is super fast on modern systems), or out of some kind of consistency concern (but this is multi-threaded, so that seems unlikely).

Possibly we should just do the full check on every query, and if that's too much we can move the `CODESEARCH_SLOWGTOD` hack into the default path with a smaller constant…